cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/services/http/handlers/WorkflowHttpHandlerTest.java

```diff
@@ -0,0 +1,755 @@
+    .registerTypeAdapter(WorkflowActionSpecification.class, new WorkflowActionSpecificationCodec())
+    .create();
+
+  private static final String PAUSE_RESUME_WORKFLOW_APP = "PauseResumeWorkflowApp";
+  private static final String PAUSE_RESUME_WORKFLOW = "PauseResumeWorkflow";
+  private static final String APP_WITH_CONCURRENT_WORKFLOW = "ConcurrentWorkflowApp";
+  private static final String APP_WITH_CONCURRENT_WORKFLOW_SCHEDULE_1 = "concurrentWorkflowSchedule1";
+  private static final String APP_WITH_CONCURRENT_WORKFLOW_SCHEDULE_2 = "concurrentWorkflowSchedule2";
+  private static final String CONCURRENT_WORKFLOW_NAME = "ConcurrentWorkflow";
+  private static final String WORKFLOW_APP_WITH_FORK = "WorkflowAppWithFork";
+  private static final String WORKFLOW_WITH_FORK = "WorkflowWithFork";
+  private static final String WORKFLOW_APP_WITH_SCOPED_PARAMETERS = "WorkflowAppWithScopedParameters";
+  private static final String WORKFLOW_APP_WITH_SCOPED_PARAMETERS_WORKFLOW = "OneWorkflow";
+  private static final String APP_WITH_SCHEDULE_APP_NAME = "AppWithSchedule";
+  private static final String APP_WITH_SCHEDULE_WORKFLOW_NAME = "SampleWorkflow";
+  private static final String APP_WITH_SCHEDULE_SCHEDULE_NAME = "SampleSchedule";
+  private static final String APP_WITH_STREAM_SCHEDULE_APP_NAME = "AppWithStreamSizeSchedule";
+  private static final String APP_WITH_STREAM_SCHEDULE_SCHEDULE_NAME_1 = "SampleSchedule1";
+  private static final String APP_WITH_STREAM_SCHEDULE_SCHEDULE_NAME_2 = "SampleSchedule2";
+  private static final String APP_WITH_STREAM_SCHEDULE_WORKFLOW_NAME = "SampleWorkflow";
+  private static final String APP_WITH_STREAM_SCHEDULE_STREAM_NAME = "stream";
+  private static final String WORKFLOW_APP_WITH_ERROR_RUNS = "WorkflowAppWithErrorRuns";
+  private static final String WORKFLOW_WITH_ERROR_RUNS = "WorkflowWithErrorRuns";
+  private static final String WORKFLOW_WITH_ERROR_RUNS_SCHEDULE = "SampleSchedule";
+
+  protected static final Type LIST_WORKFLOWACTIONNODE_TYPE = new TypeToken<List<WorkflowActionNode>>()
+  { }.getType();
+
+  private String getRunsUrl(String namespace, String appName, String workflow, String status) {
+    String runsUrl = String.format("apps/%s/workflows/%s/runs?status=%s", appName, workflow, status);
+    return getVersionedAPIPath(runsUrl, Constants.Gateway.API_VERSION_3_TOKEN, namespace);
+  }
+
+  private void checkCurrentRuns(int retries, String url, int currentRunningProgramsExpected) throws Exception {
+    int trial = 0;
+    String json;
+    List<WorkflowActionNode> output = null;
+    HttpResponse response;
+    while (trial++ < retries) {
+      response = doGet(url);
+      if (response.getStatusLine().getStatusCode() == 200) {
+        json = EntityUtils.toString(response.getEntity());
+        output = GSON.fromJson(json, LIST_WORKFLOWACTIONNODE_TYPE);
+        if (output.size() == currentRunningProgramsExpected) {
+          break;
+        }
+      }
+      TimeUnit.SECONDS.sleep(1);
```

Comment:
Don't sleep 1 second when polling in unit-test. Sleep only 50-100ms. Also, use `Tasks.waitFor()` instead, don't do custom polling loop everywhere.